### ISSUE TITLE
PVC-based OSD forgets it should be encrypted

### ISSUE THREAD

**Issue by bootc** [rook/rook#11972]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
OSD unexpectedly re-configured and "forgets" it should be encrypted.

**Expected behavior:**
When an OSD updates it should retain the knowledge that its underlying disks are (and should remain) encrypted.

**How to reproduce it (minimal and precise):**
I have no minimal reproducer for this.

The cluster has been reshaping and recovering for a while for a number of reasons. It's a pretty big cluster overall, and any changes take ages due to the amount of data involved.

There's a subset of 4x HDD nodes which have 24x HDDs and 12x SSDs. Because we couldn't configure OSDs to use half of each SSD as metadata using the usual `CephCluster.spec.storage.nodes` config we ended up adding the HDDs and SSDs as separate OSDs in different device classes. It turned out the HDDs really couldn't keep up by themselves and we needed the SSDs for metadata.

So instead we configured TopoLVM, put all the HDDs on each node in one VG and all the SSDs in another VG, then used `CephCluster.spec.storage.storageClassDeviceSets` instead, with PVCs sized such that each OSD uses a whole HDD and half of one SSD. On the whole this works surprisingly well, even with encryption which is a client requirement.

However in order for the PVC OSD prepare job to work the cluster needs to run Ceph 17.2.3 because of https://github.com/rook/rook/issues/11304 or the jobs fail. However we can't run 17.2.3 normally because https://tracker.ceph.com/issues/56530 causes the Ceph nodes to burn CPU and go really slowly. The workaround I found was to downgrade to 17.2.3 just for the prepare stage, then upgrade to 17.2.5 immediately after so the OSDs can recover/backfill correctly.

This unfortunately leads to Rook being in an almost perpetual upgrade loop, as some OSDs can't be safely shut down at all at the moment due to the recovery process and some poor decisions when initially creating some pools. We're almost done fixing that now, but at the moment the operator does loop around trying to restart most OSD pods trying to upgrade them for well over *a week* while the recovery continues.

After a few weeks of this all working fine I started work on Wednesday to find one OSD (of 72 identical PVC-backed HDD/SSD mix) stuck in `Init:CrashLoopBackOff` on the `activate` container. After much digging I realised it had lost its `encryption-open` / `encryption-open-metadata` / `blkdevmapper-encryption` / `blkdevmapper-metadata-encryption` / `encrypted-block-status` initContainers. This OSD actually made one PG go `offline` so I hackishly manually edited the Deployment and put those containers back by copying from a different OSD and substituting the relevant values. Thankfully it worked and the OSD is back in action.

Then this morning I found that exactly the same thing had happened to a different OSD on a different node overnight as well. This time there are no offline PGs so I can afford to leave it for a bit longer.

I restarted the Operator, hoping that it might fix the OSD by itself, but it doesn't seem to have made any difference.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  cephVersion:
    # https://quay.io/repository/ceph/ceph?tab=tags
    image: quay.io/ceph/ceph:v17.2.5-20230209
    # image: quay.io/ceph/ceph:v17.2.3-20220805
    # Can't use 17.2.4 or 17.2.5 when (re-)creating OSDs on ceph-hdd hosts due
    # to a Ceph bug:
    # https://github.com/rook/rook/issues/11304
    # However it's best to switch back to 17.2.5 after creation because of
    # insane CPU usage during recovery otherwise:
    # https://tracker.ceph.com/issues/56530

  dataDirHostPath: /var/lib/ceph

  mon:
    count: 3
    allowMultiplePerNode: false

  mgr:
    count: 2
    allowMultiplePerNode: false

  dashboard:
    enabled: true

  monitoring:
    enabled: true

  network:
    provider: host

  # redacted crashCollector.daysToRetain, placement, resources, priorityClassNames

  storage:
    useAllNodes: false
    useAllDevices: false

    config:
      storeType: bluestore
      encryptedDevice: "true"

    nodes:
      # redacted some all-NVMe nodes
      # here is one as-yet-unconverted to PVC HDD node:
    - name: ceph-hdd4
      deviceFilter: "^sd"

    storageClassDeviceSets:
    - name: ceph-hdd1
      count: 24

      # redacted resources, placement

      config:
        storeType: bluestore

      volumeClaimTemplates:
      - metadata:
          name: data
        spec:
          resources:
            requests:
              storage: 16763Gi
          storageClassName: lvm-hdd
          volumeMode: Block
          accessModes:
          - ReadWriteOnce
      - metadata:
          name: metadata
        spec:
          resources:
            requests:
              storage: 223Gi # floor(447.13/2)
          storageClassName: lvm-ssd
          volumeMode: Block
          accessModes:
          - ReadWriteOnce

      portable: false
      encrypted: true

    - name: ceph-hdd2
      count: 24
      # same as above but with different placement
    - name: ceph-hdd3
      count: 24
      # same as above but with different placement

  disruptionManagement:
    managePodBudgets: true

  # https://www.rook.io/docs/rook/v1.10/Storage-Configuration/Advanced/ceph-mon-health/#failing-over-a-monitor
  healthCheck:
    daemonHealth:
      mon:
        disabled: false
        interval: 45s
        timeout: 30m # default 10m
```

**Logs to submit**:

* Operator's logs, if necessary

```
2023-03-23T00:23:33Z {} 2023-03-23 00:23:33.544909 I | op-osd: updating OSD 115 on PVC "ceph-hdd2-data-14q7659"
2023-03-23T00:23:33Z {} 2023-03-23 00:23:33.544941 I | op-osd: OSD will have its main bluestore block on "ceph-hdd2-data-14q7659" and its metadata device on "ceph-hdd2-metadata-148ntmf"
2023-03-23T00:23:57Z {} 2023-03-23 00:23:57.517955 I | clusterdisruption-controller: osd "rook-ceph-osd-115" is down but no node drain is detected
```

That's basically all we get from the operator when it decides to update the OSD, removing the encryption containers.

* Crashing pod(s) logs, if necessary

```
failed to read label for /var/lib/ceph/osd/ceph-115/block: (2) No such file or directory
```

That's all we get from the failing `activate` container, but this happens because it's looking at the LUKS-encrypted block device rather than the unlocked dm-crypt device.

**Cluster Status to submit**:

* Output of krew commands, if necessary

As discussed above, we are mid-reshaping things so it doesn't look too healthy:

```
HEALTH_WARN Degraded data redundancy: 1124933/548025074 objects degraded (0.205%), 98 pgs degraded, 98 pgs undersized; 223 pgs not deep-scrubbed in time; 97 pgs not scrubbed in time
```

```
  cluster:
    id:     1e2b3821-54d6-4501-b1c7-cd342d55bbd0
    health: HEALTH_WARN
            Degraded data redundancy: 1123774/548024996 objects degraded (0.205%), 98 pgs degraded, 98 pgs undersized
            223 pgs not deep-scrubbed in time
            97 pgs not scrubbed in time
 
  services:
    mon: 3 daemons, quorum a,b,c (age 28h)
    mgr: a(active, since 29h), standbys: b
    mds: 20/20 daemons up, 20 hot standby
    osd: 176 osds: 175 up (since 16h), 175 in (since 16h); 188 remapped pgs
    rgw: 3 daemons active (2 hosts, 1 zones)
 
  data:
    volumes: 8/8 healthy
    pools:   34 pools, 4313 pgs
    objects: 103.49M objects, 284 TiB
    usage:   409 TiB used, 1.3 PiB / 1.7 PiB avail
    pgs:     1123774/548024996 objects degraded (0.205%)
             1447476/548024996 objects misplaced (0.264%)
             4113 active+clean
             98   active+undersized+degraded+remapped+backfilling
             90   active+remapped+backfilling
             12   active+clean+scrubbing+deep
 
  io:
    client:   141 MiB/s rd, 49 MiB/s wr, 790 op/s rd, 130 op/s wr
    recovery: 160 MiB/s, 43 objects/s
 ```

**Environment**:
* OS (e.g. from /etc/os-release): Debian 11 (bullseye)
* Kernel (e.g. `uname -a`): Linux admin1 6.0.0-0.deb11.6-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.0.12-1~bpo11+1 (2022-12-19) x86_64 GNU/Linux
* Cloud provider or hardware configuration: On-prem
* Rook version (use `rook version` inside of a Rook Pod): rook: v1.10.12 / go: go1.19.6
* Storage backend version (e.g. for ceph do `ceph -v`): ceph version 17.2.5 (98318ae89f1a893a6ded3a640405cdbb33e08757) quincy (stable)
* Kubernetes version (use `kubectl version`): v1.24.11+rke2r1
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): RKE2
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): see above

**Link to slack thread**:
https://rook-io.slack.com/archives/CK9CF5H2R/p1679480234209539

**Comment by travisn** [comment]
> After much digging I realised it had lost its `encryption-open` / `encryption-open-metadata` / `blkdevmapper-encryption` / `blkdevmapper-metadata-encryption` / `encrypted-block-status initContainers`.

Since this happened on the latest v1.10 release, looking in that branch, those encryption init containers are added to the OSD in [this code path](https://github.com/rook/rook/blob/release-1.10/pkg/operator/ceph/cluster/osd/spec.go#L532-L541).
```golang
if osdProps.encrypted {
	// Open the encrypted disk
	initContainers = append(initContainers, c.getPVCEncryptionOpenInitContainerActivate(osdDataDirPath, osdProps)...)
...
}
```

The `osdProps.encrypted` is set originally when the device sets are being processed from the cluster CR. As we see [here](https://github.com/rook/rook/blob/release-1.10/pkg/operator/ceph/cluster/osd/osd.go#L405), the `encrypted` property of the `osdProperties` struct is set from the `deviceSet.Encrypted` setting. 

The `deviceSet` settings can then be traced back to being set in [deviceSet.go](https://github.com/rook/rook/blob/release-1.10/pkg/operator/ceph/cluster/osd/deviceSet.go#L198) from `newDeviceSet`.

The `newDeviceSet` is a parameter that in the parent method is traced back to this [for loop](https://github.com/rook/rook/blob/release-1.10/pkg/operator/ceph/cluster/osd/deviceSet.go#L90):
```
for _, deviceSet := range c.spec.Storage.StorageClassDeviceSets {
```

Which means the `encrypted` setting came directly from the cephcluster CR settings for the storageClassDeviceSet that defined the OSD. 

But if the `encrypted` setting in the cephcluster CR changed, all of the OSDs should have been seen to have the same problem. Or at least all OSDs defined in the same `storageClassDeviceSet` should have had the same problem. 

@bootc 
- Are the two affected OSDs from the same `storageClassDeviceSet`? 
- How many OSDs are there in each device set? 
- Any message in the rook operator log that show that it's blocked from updating the OSDs because the PGs are down? Perhaps more OSDs would be affected if the update is allowed to continue, but by default `skipUpgradeChecks` could be preventing the update.

@satoru-takeuchi Do you see any other possible code path for how this could happen?

**Comment by bootc** [comment]
>     * Are the two affected OSDs from the same `storageClassDeviceSet`?

Yes, interestingly they are from the same `storageClassDeviceSet`...

🤦 now I see when I added the 3rd device set I missed the bit at the bottom that enabled this:

```
      portable: false
      encrypted: true
```

And I inserted the bit I copied *above* these two lines, so the new 3rd device set was created correctly. Indeed, only as OSDs started healing were they updated to remove the `encrypted` setting. This only happened to the 2 OSDs so far.

... but why was an OSD updated that ended up taking a PG completely offline?

>     * How many OSDs are there in each device set?

There are 24 OSDs per device set, one per host with 24 HDDs.

>     * Any message in the rook operator log that show that it's blocked from updating the OSDs because the PGs are down? Perhaps more OSDs would be affected if the update is allowed to continue, but by default `skipUpgradeChecks` could be preventing the update.

Yes as I mentioned the cluster is perpetually trying to update OSDs but can't because it's not safe to do so. Indeed you're right, it probably would have broken all the OSDs on this host (in the device set) given enough time.

I'm so sorry for wasting your time! Thanks for digging into this.
